### ISSUE TITLE
Termux doesn't terminate ssh connection on exit

### ISSUE THREAD

**Issue by trygveaa** [termux/termux-app#935]
If I ssh to a machine, and then press Exit in the notification, the ssh process on the server keeps running. This is especially visible when exiting an attached tmux session, since that session will stay attached which causes the viewport of other sessions to be reduced to the size of the phone.

Note that this happens on my new Pixel 3, but did not happen on my previous Nexus 4. They both have version 0.65, so I'm not sure what's the difference.

**Comment by Neo-Oli** [comment]
Pressing the Exit button will only kill all sessions. It will not kill any background processes. Are you perhaps running your ssh process in a tmux session, because the tmux server survives the "Exit".

**Comment by trygveaa** [comment]
I'm not running tmux on my phone, only on the server. On the phone, I'm just using the ssh command.

I tried running ps after exiting Termux with an ssh-session, and I can see that the ssh command is still running.

**Comment by xeffyr** [comment]
I tried the following:
1. Ssh to remote server
2. Launch tmux on remote server
3. Tap on "exit" button to close termux app.

\- then `ps aux`. As result, *issue wasn't reproduced*. Even `pidof ssh` didn't show anything. Session was terminated successfully.

**Comment by trygveaa** [comment]
@xeffyr: Yeah, it's only a problem on some phones/OS versions apparently. I didn't have this problem on my Nexus 4 or Nexus 5X with Android 8, but it is a problem on Pixel 3 with Android 9.

**Comment by lotheac** [comment]
This is not specific to ssh, I was able to reproduce with just `cat`. Started seeing this today on my Nokia 8 after upgrading it from Android8 to Android9.

```
$ ps -o tty,pid,comm
TT     PID   COMMAND
?      23551 cat
136,1  23605 bash
136,1  23754 ps
```

`cat` doesn't seem to have a tty any more (terminating the termux session with the exit button took it away), but it's still running.

**Comment by xeffyr** [comment]
> terminating the termux session with the exit button took it away), but it's still running.

@lotheac It is terminates correctly when you hit "exit" button or it still running after launching Termux app again ?

**Comment by lotheac** [comment]
@xeffyr sorry, I was unclear. Repro steps for me are:

1. start termux (no sessions running)
2. run `cat` in this new tty
3. go to notification to hit the Exit button. termux notification disappears
4. open a new termux session, and run `ps -o tty,pid,comm`. Observe `cat` is still there.

**Comment by xeffyr** [comment]
> cat doesn't seem to have a tty any more

This happens only with "cat" or with all programs (e.g. bash) from terminated session ?

**Comment by lotheac** [comment]
Well, it happened with ssh for me first (as this issue describes), I just thought to try `cat` as another program. The parent shell for `cat` does not stick around (and I cannot reproduce the issue with another `bash` as a child of the initial one)

**Comment by lotheac** [comment]
actually it doesn't repro every time, so maybe this is a race somewhere.

**Comment by trygveaa** [comment]
Yes, it happens with other programs, e.g. `cat` and `sleep` for me too. I guess the reason it doesn't happen with `bash` is that `bash` probably exits when it looses the terminal, since it's an interactive process.

I've never experienced it not happening, it seems to happen every time for me.

If I close Termux from the app switcher after I press Exit from the notifications, the processes are killed, but not before. I.e. I have to first press Exit, then close Termux from the app drawer. If I only do one of those things, the processes persist.

**Comment by xeffyr** [comment]
If bash exits, then all it's child processes should exit too. This happens on Android 8 and lower as it should. But on Android 9 this mechanism seems broken.

**Comment by lotheac** [comment]
But when you hit the 'Exit' key in the notification, and then look at the task switcher's screenshot of the process, you can see text `[Process completed (signal 9 - press Enter]`. So something is delivering SIGKILL.

**Comment by lotheac** [comment]
It's another matter though why the `cat` doesn't subsequently get an error for its blocking read() on stdin, if the tty stops existing. Could very well be a bug in Android.

**Comment by xeffyr** [comment]
> So something is delivering SIGKILL.

Yeah, shell receives SIGKILL.  That's okay.
``` .java
    public void finishIfRunning() {
        if (isRunning()) {
            try {
                Os.kill(mShellPid, OsConstants.SIGKILL);
            } catch (ErrnoException e) {
                Log.w("termux", "Failed sending SIGKILL: " + e.getMessage());
            }
        }
    }
```

I'll try to manually send a kill signal to the shell without closing session to see what happens.

**Comment by trygveaa** [comment]
Huh, why does Termux send SIGKILL? Shouldn't the SIGHUP you mentioned suffice? That explains why my bash history hasn't always been saved (separate issue though).

> Running killall -9 bash terminates all processes including running "cat".

This also happens for me, but if I run `sleep 100` instead of `cat`, it persists.

**Comment by xeffyr** [comment]
> Huh, why does Termux send SIGKILL?

If I replace SIGKILL with SIGHUP, nothing changes. Bash stopped as it should, but "cat" left in background.

> Shouldn't the SIGHUP you mentioned suffice?

Yes, should be enough. At least, it will allow shell to save history.
I'll look on replacing kill signal with hup, if no bad effects a pull request will be submitted.

Regarding this issue, it may come from Android. But I don't know why.
Have you checked if same issue reproducible in https://play.google.com/store/apps/details?id=jackpal.androidterm ?

**Comment by xeffyr** [comment]
With [jackpal's androidterm](https://play.google.com/store/apps/details?id=jackpal.androidterm ?) things even worse than with Termux - shells are not terminated. It's definetly Android issue.

**Comment by trygveaa** [comment]
> If I replace SIGKILL with SIGHUP, nothing changes. Bash stopped as it should, but "cat" left in background.

Yes, same for me. It's unrelated to this issue, so sorry for bringing it up here.

> I'll look on replacing kill signal with hup, if no bad effects a pull request will be submitted.

Thanks

> With [jackpal's androidterm](https://play.google.com/store/apps/details?id=jackpal.androidterm) things even worse than with Termux - shells are not terminated

How did you test this? I tried, but `ps` in that app only shows me processes from the same window, so I can't see if the processes are still running.

**Comment by xeffyr** [comment]
> I tried, but ps in that app only shows me processes from the same window

`ps -A -o tty,comm`.

**Comment by trygveaa** [comment]
Ah, thanks. Yeah, they persist here too with that app.

**Comment by pvonmoradi** [comment]
@trygveaa Is there a way to `pkill` all bg processes upon exiting all Termux instances?

**Comment by xeffyr** [comment]
@pvonmoradi Termux sends SIGKILL to processes on exit.

**Comment by pvonmoradi** [comment]
@xeffyr I wonder if it is intended behavior...Personally, I feel like termux initiated daemons should be killed upon exiting termux. For example I run `sshd` but even after quitting, I can still ssh into the device which is a little weird.  
(Force stopping Termux causes it to be killed)

**Comment by trygveaa** [comment]
@xeffyr: This issue is about foreground processes remaining after exit though, so apparently something doesn't work, unless it's been fixed?

To avoid the issue, I have just stopped using the exit button in the notification (also because the notification started appearing further down in the list), and use buttons on the extra keys row to exit bash and tmux instead.

**Comment by pvonmoradi** [comment]
@trygveaa executing `exit` or `ctrl+d` or `Exit from Termux notifications` does not kill the `sshd` in my case.

**Comment by trygveaa** [comment]
@pvonmoradi: As I said this issue is about foreground processes, not background processes. At least that's what I was talking about when I opened the issue.

**Comment by xeffyr** [comment]
> unless it's been fixed?

@trygveaa I just reproduced the issue on Pixel 5, with `ssh` session to remote `tmux`. So it is not fixed. But I can't reproduce it with something else, e.g. running `htop` is being killed on exit as it should be.

***

> executing exit or ctrl+d or Exit from Termux notifications does not kill the sshd in my case.

This shouldn't kill `sshd`. That's behavior of all terminal emulators (at least on normal Linux distributions), not just Termux. Daemonized processes do not belong to current session and are not attached to terminal.

**Comment by trygveaa** [comment]
> But I can't reproduce it with something else, e.g. running htop is being killed on exit as it should be.

I assume `htop` and most other interactive processes terminate themselves when they loose the terminal. E.g. `sleep` remains for me, just like `ssh`.

**Comment by thet83** [comment]
I 'v noticed this as well but i think it's actually quite easy to "workaround" using this command to start sshd:
`sshd -D &`

and placing the following in .bashrc
`trap 'kill $(jobs -p)' EXIT`

**Comment by thet83** [comment]
I'v noticed this as well, but found a "workaround"
if you start sshd like this:
`**sshd -D**`
and place this line in .bashrc
`**trap 'kill $(jobs -p)' EXIT**`
it will make shure that sshd will exit before termux is closed
the only "problem" with this is, that sshd seems to take some time to shut down, so termux may take a while before its closed, but if testet a few times and i think it's working like this

**Comment by TukangM** [comment]
just restart your devices and the sshd terminated.

maybe the problem with custom kernel

**Comment by defkev** [comment]
@thet83's observation still applies to termux 0.118.0 on Android 11

* Start termux
* `sshd -D` (don't daemonize)
* Hit Exit on termux notification
* Start termux
* `sshd -D`

Fails with bind error, port in use.

`pstree -p` lists the orphant process, `pkill sshd` to get it working again.

>just restart your devices and the sshd terminated.

This isn't Windows
